Adding declarations to llvmcall

vchuravy opened the pull request:
Rebased from #8740 via #11516 in response to #11603 

This allows to hook up llvm intrinsics via llvmcall. This is necessary for the CUDA backend work and several people expressed that they have other usages in mind. 

Ping @ihnorton @tkelman 
## Overview

This extension of llvmcall allows to pass declarations of intrinsics to llvmcall. As an example see:

``` julia
function ceil(x::Float64)
    llvmcall(
        ("declare double @llvm.ceil.f64(double)",
         """%2 = call double @llvm.ceil.f64(double %0)
            ret double %2"""),
    Float64, Tuple{Float64}, x)
end

function ceilfloor(x::Float64)
    llvmcall(
        ("""declare double @llvm.ceil.f64(double)
            declare double @llvm.floor.f64(double)""",
         """%2 = call double @llvm.ceil.f64(double %0)
            %3 = call double @llvm.floor.f64(double %2)
            ret double %3"""),
    Float64, Tuple{Float64}, x)
end
```

Commit:
Extend llvmcall with support for entering declarations.

vchuravy committed:
adjust llvmcall test to tuple changes

tkelman commented on test/llvmcall.jl:
```diff
@@ -48,3 +48,44 @@ baremodule PlusTest
+    llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
+        ret double %2""", Float64, Tuple{Float64}, x)
+end
+@test_throws ErrorException undeclared_ceil(4.2)
```
This is apparently failing - does this functionality only work with LLVM 3.5+?

vchuravy replied:
The test seems to be quite finicky

Executing from the REPL:
````
julia> function undeclared_ceil(x::Float64)
           Base.llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
               ret double %2""", Float64, Tuple{Float64}, x)
       end
undeclared_ceil (generic function with 1 method)

julia> undeclared_ceil(4.2)
ERROR: error compiling undeclared_ceil: Failed to parse LLVM Assembly: 
julia: <string>:3:18: error: use of undefined value '@llvm.ceil.f64'
%2 = call double @llvm.ceil.f64(double %0)
````

````
try 
         undeclared_ceil(4.3)
       catch err
         println(err)
       end
ErrorException("error compiling undeclared_ceil: Failed to parse LLVM Assembly: \njulia: <string>:3:18: error: use of undefined value '@llvm.ceil.f64'\n%2 = call double @llvm.ceil.f64(double %0)\n                 ^\n")
````

````
julia> Test.@test_throws ErrorException undeclared_ceil(4.3)

julia> Test.@test_throws ErrorException undeclared_ceil(4.3)
ERROR: test failed: undeclared_ceil(4.3) did not throw ErrorException
 in expression: undeclared_ceil(4.3)
 in error at error.jl:21
 in default_handler at test.jl:29
 in do_test_throws at test.jl:74
````

````
julia> try 
         undeclared_ceil(4.3)
       catch err
         println(typeof(err))
       end
ErrorException
````

tkelman replied:
Yes I noticed similar unpredictable behavior. Maybe the GC root will help?

yuyichao replied:
That I really doubt... But maybe.....
The GC root needs to be fix though, unless I'm missing sth.

yuyichao commented on src/ccall.cpp:
```diff
@@ -600,10 +604,20 @@ static Value *emit_llvmcall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+        // if the IR is a tuple, we expect (declarations, ir)
+        if (jl_nfields(ir) != 2)
+            jl_error("Tuple as first argument to llvmcall must have exactly two children");
+        decl = jl_fieldref(ir,0);
```
Missing GC root? (`ir` is overriden below and (new) line `628` is allocating)

vchuravy committed:
add missing gc root

vchuravy commented:
There is one problem that I see and I don't for certain how to best solve it. Looking at the test failure wrt `undeclared_ceil`. One the first call to the function I get the expected error, but on the second one I get a valid return value:

````
julia> function undeclared_ceil(x::Float64)
           Base.llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
               ret double %2""", Float64, Tuple{Float64}, x)
       end
undeclared_ceil (generic function with 1 method)

julia> undeclared_ceil(4.2)
ERROR: error compiling undeclared_ceil: Failed to parse LLVM Assembly: 
julia: <string>:3:18: error: use of undefined value '@llvm.ceil.f64'
%2 = call double @llvm.ceil.f64(double %0)
                 ^


julia> undeclared_ceil(4.2)
5.0
````

Even better calling a function with an undefined intrinsic throws the error and then adding a second function that also calls the intrinsic does not throws an error for the second function"

````
julia> function undeclared_ceil(x::Float64)
           Base.llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
               ret double %2""", Float64, Tuple{Float64}, x)
       end
undeclared_ceil (generic function with 1 method)

julia> undeclared_ceil(4.1)
ERROR: error compiling undeclared_ceil: Failed to parse LLVM Assembly: 
julia: <string>:3:18: error: use of undefined value '@llvm.ceil.f64'
%2 = call double @llvm.ceil.f64(double %0)
                 ^


julia> function undeclared_ceil2(x::Float64)
           Base.llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
               ret double %2""", Float64, Tuple{Float64}, x)
       end
undeclared_ceil2 (generic function with 1 method)

julia> undeclared_ceil2(4.1)
5.0
````

vchuravy commented:
@maleadt added a comment about that case:

````
# KNOWN ISSUE: although the llvmcall in undeclare_ceil did error(), the LLVM
# module contains the generated IR referencing the ceil intrinsic. At some
# point, a declaration for that intrinsic is added, breaking any future
# llvmcall((conflicting declaration, ...), ...). Not an issue in LLVM 3.5+.
````

Suggestions are welcome

tkelman commented on src/ccall.cpp:
```diff
@@ -600,10 +604,22 @@ static Value *emit_llvmcall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+            jl_error("Tuple as first argument to llvmcall must have exactly two children");
+        decl = jl_fieldref(ir,0);
+        ir = jl_fieldref(ir,1);
+	// Need GC root because of allocating call later
```
spaces instead of tabs

tkelman commented:
I don't know whether there's any downside to enabling this feature, but if it would be useful to have now despite the tests doing funky things on LLVM 3.3, could we just comment out the unpredictable tests for now, and revisit after upgrading LLVM? #9336

vchuravy committed:
llvmcall test is not working on llvm 3.3

The test that checks llvm behaviour for undefined intrinsics does not
work on llvm 3.3. It works(tm) under llvm 3.5+.

vchuravy commented:
I moved the gc root to the right place (hopefully) and commented out the wonky test.

yuyichao commented:
@vchuravy Yup the GC root looks good to me now =).

vchuravy commented:
Most of the added tests check the behaviour of llvm with regards to undefined or double defined intrinsics declarations, so for the time being I commented them out.

vchuravy commented:
Instead of out-commenting I decided to future proof the test by adding a conditional on them. So when #9336 occurs we don't forget about the tests.

vchuravy committed:
prevent adding a declare that already exists in the module

getNamedValue expects getNamedValue("llvm.floor.f64") for "declare
double @llvm.floor.f64(double)"

vchuravy commented:
The Travis error on 64bit looks like a timeout to me, the llvmcall tests passed (Hurray). All test pass locally.

vchuravy committed:
parse multiline declarations in llvmcall

vchuravy commented:
So I think this is round for another look from somebody else. The last commit extends llvmcall to take  multiple lines of declarations

SimonDanisch commented:
Compilation works, but there are test errors:
```
exception on 5:
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: UNKNOWN at 0x7ff981ff8b9c -- RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
Unwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
_cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
showerror at replutil.jl:91
jlcall_showerror_26418 at  (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at client.jl:88
jl_f_apply at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
with_output_color at util.jl:331
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
display_error at client.jl:86
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
run_work_thunk at multi.jl:608
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6aa0abc1 -- utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_svec_fill at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_gc_run_all_finalizers at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_atexit_hook at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_exit at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_throw_in_ctx at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
_chkstk at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
Unwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
_cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
showerror at replutil.jl:91
jlcall_showerror_26418 at  (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at client.jl:88
jl_f_apply at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
with_output_color at util.jl:331
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
display_error at client.jl:86
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
run_work_thunk at multi.jl:608
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
        From worker 5:       * linalg/triangular   Worker 5 terminated.
ERROR (unhandled task failure): readcb: connection reset by peer (ECONNRESET)
 in remotecall_fetch at multi.jl:688
 in remotecall_fetch at multi.jl:693
 in anonymous at task.jl:1423
        From worker 3:       * linalg/cholesky      in 149.97 seconds
        From worker 4:       * linalg/lu            in 150.09 seconds
        From worker 2:       * linalg1              in 333.50 seconds
ERROR: LoadError: ProcessExitedException()
 in remotecall_fetch at multi.jl:688
 in remotecall_fetch at multi.jl:693
 in anonymous at task.jl:1423
while loading C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\share\julia\test\runtests.jl, in expression starting on line 5
```

tkelman commented:
@SimonDanisch what's your `versioninfo()` there? How was that built?

SimonDanisch commented:
Ah yeah, that was not very informative.
Windows + cygwin. Checked out JuliaGPU->vc/llvmcall + make (from fresh git clone).
```
Julia Version 0.4.0-dev+5241
Commit f8ce3bd (2015-06-08 14:37 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```
Do I need LLVM 3.5, or what?

SimonDanisch commented:
and I did `make win-extras` + `make binary-dist` and then installed it.

tkelman commented:
Thanks, I'll see if I can reproduce. Might be Haswell-specific?

vchuravy commented:
@SimonDanisch could you check out dff768a14dd0f4640451699a8ec25a714ed63b46 and see if the error is persistent there? That is the commit from which `vc/llvmcall` is branched off.

I would like to add some documentation for `llvmcall` In which section should that best go?
@Keno, since you wrote the original version, do you have some ideas on documentation?

tkelman commented:
ref #10760

SimonDanisch commented:
```
julia> versioninfo()
Julia Version 0.4.0-dev+5235
Commit dff768a* (2015-06-06 12:56 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```
Looks quite similar:
```
C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin>julia.exe C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\share\julia\test\runtests.jl al
        From worker 5:       * linalg4              in  17.21 seconds
        From worker 5:       * linalg/lapack        in  24.04 seconds
        From worker 4:       * linalg3              in  48.03 seconds
        From worker 4:       * linalg/tridiag       in   9.33 seconds
        From worker 3:       * linalg2              in  64.73 seconds
        From worker 4:       * linalg/bidiag        in  20.47 seconds
        From worker 3:       * linalg/diagonal      in  20.77 seconds
        From worker 3:       * linalg/givens        in   2.71 seconds
        From worker 4:       * linalg/pinv          in  11.78 seconds
        From worker 3:       * linalg/cholesky      in  45.25 seconds
        From worker 4:       * linalg/lu            in  44.96 seconds
        From worker 2:       * linalg1              in 138.16 seconds
        From worker 3:       * linalg/symmetric     in   9.31 seconds
        From worker 3:       * keywordargs          in   2.12 seconds
exception on 4:
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: UNKNOWN at 0x7fff2d268b9c -- RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
UNKNOWNUnwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
 at 0x7fff2d268b9c -- _cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
Unwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
_cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
jl_static_eval at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
jl_static_eval at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_compile at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
do_test at test.jl:49
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_interpret_toplevel_expr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_interpret_toplevel_thunk_with at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
display_error at client.jl:86
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_eval_with_compiler_p at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
run_work_thunk at multi.jl:608
jl_parse_eval_all at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_ at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
unknown function (ip: 1691279766)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
runtests at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\share\julia\test\testdefs.jl:197
jlcall_runtests_21263 at  (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_f_apply at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at multi.jl:854
run_work_thunk at multi.jl:605
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6b05b527 -- utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6b49194c -- utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_dump_llvm_value at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_static_eval at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_compile at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_svec_fill at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_gc_run_all_finalizers at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_atexit_hook at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_exit at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_throw_in_ctx at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
_chkstk at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
Unwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
_cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_svec_fill at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_gc_run_all_finalizers at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_static_eval at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_static_eval at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
        From worker 2:       * core                jl_atexit_hook at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (u
jl_exit at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_throw_in_ctx at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
_chkstk at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RtlRaiseException at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
RaiseException at C:\WINDOWS\system32\KERNELBASE.dll (unknown line)
Unwind_RaiseException at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libgcc_s_seh-1.dll (unknown line)
_cxa_throw at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
Znwy at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libstdc++-6.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
utf8proc_NFKC at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_generate_fptr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
display_error at client.jl:86
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
run_work_thunk at multi.jl:608
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_and_lookup at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_compile at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_trampoline at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
do_test at test.jl:49
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_interpret_toplevel_expr at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_interpret_toplevel_thunk_with at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_eval_with_compiler_p at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_parse_eval_all at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_load_ at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
unknown function (ip: 1691279766)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
runtests at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\share\julia\test\testdefs.jl:197
jlcall_runtests_21263 at  (unknown line)
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
jl_f_apply at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at multi.jl:854
run_work_thunk at multi.jl:605
jl_apply_generic at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
anonymous at task.jl:854
jl_unprotect_stack at C:\Users\Sim\AppData\Local\Julia-0.4.0-gpu\bin\libjulia.dll (unknown line)
Worker 2 terminated.ERROR (unhandled task failure): InexactError()
 in _uv_hook_return_spawn at process.jl:238
        From worker 5:       * linalg/triangular    in 131.21 seconds
```

vchuravy commented:
@SimonDanisch thanks for checking that :) That means it is not the fault of this PR. Could you check against current master and if it persists open an issue about it? If it goes away fixing it here should be then as simple as a rebase.

SimonDanisch commented:
I'm on it ;) compilation is just always a little slow...

SimonDanisch commented:
Okay... So it first gave me an OutOfMemoryException(), but after a restart with fewer processes master completed successfully.
```
Julia Version 0.4.0-dev+5293
Commit 608fc91* (2015-06-10 09:21 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

vchuravy commented:
So I was wondering whether or not we can get this into `0.4` or if we want to wait for `0.5`.

ihnorton commented on src/ccall.cpp:
```diff
@@ -691,6 +704,35 @@ static Value *emit_llvmcall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+                if (llvmcallDecls.count(declhash) == 0) {
+                    // Find name of declaration
+                    char *declcopy = (char*) malloc(strlen(declstr) + 1);
+                    strcpy(declcopy, declstr);
```
Since this is C++, could use std::string::find and substr here.

vchuravy commented:
I am going to wait till #11973 is merged and then rebase this PR.

vchuravy committed:
use std::string tools for declaration name extraction

vchuravy committed:
update test comments

vchuravy commented:
Rebased and updated this PR

ping @Keno @ihnorton @tkelman @yuyichao 
This should be ready for another round of reviews.

vchuravy commented:
@Keno Is this the right approach or should one rather use `Cxx.jl` to create a binding to llvm and call the C++ api?

hayd commented:
Bump! This needs another rebase.

vchuravy commented:
@ViralBShah @hayd Rebased. 

Since we now have documentation for llvmcall I will write additional documentation for this.

vchuravy commented:
The failing build seems to be a homebrew issue.
```
Error: No such file or directory - /usr/local/Cellar/oniguruma/5.9.6/share/emacs/site-lisp/oniguruma
```

tkelman commented:
Looked like a download failure, restarted the build and it passed. This LGTM if other people think it's safe.

vchuravy committed:
Document declarations for llvmcall

[ci skip]

vchuravy commented:
So I added a short documentation for the declarations, I hope I got the format right.

vchuravy commented:
@Keno what do you think?

tkelman commented:
@Keno @vtjnash can we merge this?

tkelman commented:
Alright, tired of waiting for review from unresponsive code owners on this. If it causes any problems they can be addressed going forward.

vchuravy commented:
Thank you all for staying on this with me :)

tkelman commented:
#11516 has the potential to be a pretty big deal, so thank you for seeing it through and making it happen. What's next, #9423?

vchuravy commented:
Yeah that is next, but it will need a heavy rebase due to the codegen
changes.

On Sun, 15 Nov 2015 at 23:27 Tony Kelman <notifications@github.com> wrote:

> #11516 <https://github.com/JuliaLang/julia/pull/11516> has the potential
> to be a pretty big deal, so thank you for seeing it through and making it
> happen. What's next, #9423 <https://github.com/JuliaLang/julia/pull/9423>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/11604#issuecomment-156814857>.
>

maleadt commented:
Thanks for getting this merged! I only now noticed the declaration extraction, why exactly do we do that? It also doesn't seem compatible with global variable declarations (a la `@foo = bar`), but is just a matter of adjusting the parsing.

vchuravy commented:
I added the function name parsing to prevent adding already know declarations to the module, which would lead to an error (on llvm versions where Julia uses one module for everything). Since llvmcall is only meant for functions, I am unsure if adding more global state, is a good idea. What do you need global variables for? And would that work if each function has its own module?

maleadt commented:
Previously, we just added the entire declaration string into the `llvmcallDecls`, which also prevented duplicate declarations. I'm just unsure why parsing the declaration name is necessary, splitting on endline boundaries seems sufficient to me.

Concerning global variables, these just behave like declarations (only use them after being added once to the module). I'm currently using it for hackish shared memory support (`decl: @shmem = ...()`, `ir: load/store(@shmem, ...)`) but I'm sure there's more legit use cases with them.

vchuravy commented:
Hm the problem was that we would insert these declarations into a module where they already are defined. So as an example code like https://github.com/JuliaLang/julia/blob/8c23dc8816114c3b32dc737de995b7c2c645d749/test/llvmcall.jl#L96-L117 would fail, because both declarations would end up in the same llvm module.

maleadt commented:
But the declaration string for `declared_floor` equals that of `doubly_declared_floor`, so this should be caught? I'll have a stab at it, having good tests makes this easy to test :-) If not, I guess tweaking the declaration parsing should fix the global variable use case. I'll add a test for that too then.

vchuravy commented:
Yes if the declaration came in via llvmcall it will be caught, but not if it came in via one of the base intrinsics (I might be miss remembering things)

maleadt commented:
Problem is, the intrinsics defined outside of `llvmcall` will not be present in the `llvmcalldecls` set, so we can't check against them. In other words, it would still result in a double declaration.
This does not hold for the base intrinsics though, because these seem to be added through `getDeclaration(llvm::Intrinsic)`, which doesn't physically add a `declare` statement to the module.

vchuravy commented:
Yes, that's why we need to parse them so we can check if they are already present in the module https://github.com/JuliaGPU/julia/commit/2ca11b6f5d39b4bf7d36aeb157871a2d663039df#diff-be60d844ffde1319a02946d17c472e46R629

maleadt commented:
Ah completely missed that, sorry for the noise. I'll update the declaration parsing then.
